internal/server/v1/observations/calculator.go

```diff
@@ -0,0 +1,302 @@
+	list := [][]string{} // A list of lists of series date keys.
+	for _, varInfo := range c.varInfoMap {
+		dateKeys := []string{}
+		for _, series := range varInfo.seriesCandidates {
+			dateKeys = append(dateKeys, series.dateKey())
+		}
+		list = append(list, dateKeys)
+	}
+	commonDateKeys := util.StringListIntersection(list)
+	if len(commonDateKeys) == 0 {
+		return fmt.Errorf("no same date range for input time series sets")
+	}
+
+	// Choose the longest date key(s), used for selecting series among seriesCandidates.
+	// The longest date key represents the longest coverage of dates for time series.
+	longestDateKeySet := map[string]struct{}{}
+	maxDateKeyLength := 0
+	for _, k := range commonDateKeys {
+		if l := len(k); l > maxDateKeyLength {
+			for k := range longestDateKeySet {
+				delete(longestDateKeySet, k)
+			}
+			longestDateKeySet[k] = struct{}{}
+			maxDateKeyLength = l
+		} else if l == maxDateKeyLength {
+			longestDateKeySet[k] = struct{}{}
+		}
+	}
+
+	// Set chosenSeries for each varInfo.
+	for _, varInfo := range c.varInfoMap {
+		filteredSeriesCandidates := []*calculatorSeries{}
+		for _, series := range varInfo.seriesCandidates {
+			if _, ok := longestDateKeySet[series.dateKey()]; ok {
+				filteredSeriesCandidates = append(filteredSeriesCandidates, series)
+			}
+		}
+		varInfo.chosenSeries = rankCalculatorSeries(filteredSeriesCandidates)
+	}
+
+	return nil
+}
+
+// Recursively iterate through the AST and compute the result.
+//
+// The existing ast.Walk() doesn't work here, because in ast.Walk(), visiting children is the last
+// step of the function that visits parent, so the information in child nodes cannot be passed
+// above in the iteration.
+func (c *calculator) evaluateExpr(node ast.Node) (*calculatorSeries, error) {
+	// If a node is of type *ast.Ident, it is a leaf with a series value.
+	// Otherwise, it might be *ast.ParenExpr or *ast.BinaryExpr, so we continue recursing it to
+	// compute the series value for the subtree..
+	computeChildSeries := func(node ast.Node) (*calculatorSeries, error) {
+		if reflect.TypeOf(node).String() == "*ast.Ident" {
+			return c.varInfoMap[node.(*ast.Ident).Name].chosenSeries, nil
+		}
+		return c.evaluateExpr(node)
+	}
+
+	switch t := node.(type) {
+	case *ast.BinaryExpr:
+		xSeries, err := computeChildSeries(t.X)
+		if err != nil {
+			return nil, err
+		}
+		ySeries, err := computeChildSeries(t.Y)
+		if err != nil {
+			return nil, err
+		}
+		return evaluateBinaryExpr(xSeries, ySeries, t.Op)
+	case *ast.ParenExpr:
+		return c.evaluateExpr(t.X)
+	default:
+		return nil, fmt.Errorf("unsupported ast type %T", t)
+	}
+}
+
+// This implements ast.Visitor, used by ast.Walk, which iterates AST with a depth-first order.
+// For each ast.BinaryExpr, it extracts the "Name" property of X and Y.
+// For example, for an expression "Count_Person - Count_Person_Female[mm=USCensus]", it returns
+// {Count_Person, Count_Person_Female[mm=USCensus]}.
+type varNameVisitor struct {
```

Comment:
is it worth to have a file for varNameVisitor struct?

Reply:
done